### PR TITLE
Upgrades to HIPAlignmentAlgorithm functionality (part of May-July 2017 tracker alignment campaign for pixel/strips alignment)

### PR DESCRIPTION
**Pull request by usarica**
Motivation:
- The submitted jobs often required the submitter to watch them, and submission involved steps that could be automatized.
- Alignment starting from slightly different conditions (GT, extra loads to GT) required extra intermediate steps
1) Re-organized batchHippy and configureHippy
2) Implemented an option parser class
3) Changes were propagated to the templates cfg files that reside in python. Flexibility is given to change this directory, and the name of some of these files.
- Hardcoded formula for track eta-dependent reweighting of chi-squared was present in the C++ code
1) Option migrated to the cfi file, the exact formula string is kept in a commented line in align cfg.
- A tight max. relative parameter error cut was present and was found to be ineffective
1) Option defaulted to -1 everywhere
- Alignable-specific cuts were added.
- Per-alignable reweighting of chi-square based on different data samples is now supported as another way to reweight (not based on tracks but actually on hits count in the particular alignable).
- The algorithm did not support the alignment of surface deformations. This feature is now supported.
1) Additional options in the cfi file. Configuration example in common cfg in python/ (but disabled)
2) Some revision and rearrangement was needed in HIPAlignmentAlgortihm, HIPUserVariables. New container classes were also added for the alignable-specific cuts.
3) Alignment of surface deformations made optional through a flag in batchHippy.

Checks: 
- Checks were done running different alignments with the pixel and strips at module, module unit or higher level alignments as part of the (ongoing) tracker alignment campaign.
- See these slides for a quick overview of items implemented and checks done:
https://indico.cern.ch/event/624094/contributions/2627981/attachments/1476471/2287406/20170614.pdf
https://indico.cern.ch/event/624094/contributions/2627967/attachments/1476540/2287679/hpvalidation.pdf
https://indico.cern.ch/event/649295/contributions/2653665/attachments/1488080/2312139/20170705.pdf

Notes
- Rebased on top of the commits in PR #19583 (includes #19541) because these changes are essential for the HIPAlignmentAlgorithm to function properly.

### CODE REVIEW

**Comment by usarica** [comment]
Tagging @hroskes, who authored commits in this PR as well

**Comment by perrotta** [comment]
please test

**Comment by usarica** [comment]
Sorry, I noticed a typo in OptionParser. It should not affect any of the checks done though.

**Comment by slava77** [comment]
There are quite a few changes in reco comparisons, which all apparently originate in the #19541 
I posted a comment in that PR since it is most factorized.
I can imagine that the updates may have to be pushed just into this PR to avoid a triplicate update.

**Comment by usarica** [comment]
Ok, at least it is good that we factorized the PRs this way. Let me take a look if I can find something obvious in the Gsf track fitters.

**Comment by usarica** [comment]
The PR was updated based on update in #19541

**Comment by ianna** [comment]
please test

**Comment by usarica** [comment]
(Also for PR #19583)I don't think that this error has anything to do with the PR. There was no modification to material budget.

**Comment by perrotta** [comment]
@usarica : that error is likely related to the merging of #19469 in the latest IB's
There is discussion on github on that PR's thread: it looks like the issue has been identified, and a solution may be at hand.

**Comment by usarica** [comment]
@perrotta Thank you very much for clarification, Is the fact that there don't seem to be any old vs new comparison plots related to the add-on failure? I also saw this issue in #19541 (only part of the commit from here), but there was no add-on failure in that test.

**Comment by perrotta** [comment]
All comparisons from the automatically run tests are available in https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_9_3_X_2017-07-09-2300+19584/21188/, as correctly reported in https://github.com/cms-sw/cmssw/pull/19584#issuecomment-314046416

**Comment by usarica** [comment]
But then why is this plot dir. empty?
https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_9_3_X_2017-07-09-2300+19584/21188/validateJR/all_OldVSNew_RunPhoton2012Bwf4p53/

Are plots recorded only when there are differences? I do see compilation errors in the log.

**Comment by perrotta** [comment]
Ulascan Sarica <notifications@github.com> ha scritto:

> But then why is this plot dir. empty?
> https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_9_3_X_2017-07-09-2300+19584/21188/validateJR/all_OldVSNew_RunPhoton2012Bwf4p53/
>
> Are plots recorded only when there are differences?

Yes, only plots with differences are shown

> I do see compilation errors in the log.

I don't, only one warning in the job that makes the comparison

**Comment by usarica** [comment]
I see, okay then I won't worry about it.

**Comment by slava77** [comment]
+1

for https://github.com/cms-sw/cmssw/pull/19584	7455b7d
- reco changes are only in RecoTracker/TrackProducer and in TrackingTools/TrackFitters, the same as in https://github.com/cms-sw/cmssw/pull/19541	b9f5ccf; apart for the technical updates, this avoids problems in fits which apparently do not appear in the current standard workflows
- jenkins tests pass and comparisons show no differences

**Comment by ianna** [comment]
+1

**Comment by usarica** [comment]
Hi,
It looks like we are still pending a fix to the material budget error in the checks, which is not caused by the changes in the PR, and sign-off from ALCA conveners (Gregor is away these days). What would be the pathway to proceed with this PR?

**Comment by davidlange6** [comment]
please test

**Comment by arunhep** [comment]
+1

File: Alignment/CommonAlignmentAlgorithm/src/AlignmentParameterStore.cc [hunk 1]
```diff
@@ -445,53 +446,89 @@ applyAlignableAbsolutePositions( const align::Alignables& alivec,
...
-	}
+        if (found) {
+          edm::LogError("DuplicatePosition")
+            << "New positions for alignable found more than once!";
+        }
+        else {
+          // New position/rotation
+          const align::PositionType& pnew = ipos->pos();
+          const align::RotationType& rnew = ipos->rot();
+          const std::vector<double> dnew = ipos->deformationParameters();
```
Comment (davidlange6) [inline Alignment/CommonAlignmentAlgorithm/src/AlignmentParameterStore.cc]: could this be a const &?

File: Alignment/CommonAlignmentAlgorithm/src/AlignmentParameterStore.cc [hunk 2]
```diff
@@ -445,53 +446,89 @@ applyAlignableAbsolutePositions( const align::Alignables& alivec,
...
+          align::RotationType rotDiff = rold.multiplyInverse(rnew);
+          align::rectify(rotDiff); // correct for rounding errors 
+          ali->move(posDiff);
+          ali->rotateInGlobalFrame(rotDiff);
+          LogDebug("NewPosition") << "moving by:" << posDiff;
+          LogDebug("NewRotation") << "rotating by:\n" << rotDiff;
+
+          // add the surface deformations
+          // If an old surface deformation record exists, ensure that the added deformation has the same type and size.
+          if (dold.size() != 0 && dtype != SurfaceDeformationFactory::kNoDeformations && dnew.size()==dold.size()){
```
Comment (davidlange6) [inline Alignment/CommonAlignmentAlgorithm/src/AlignmentParameterStore.cc]: please use dold.empty()

File: Alignment/CommonAlignmentAlgorithm/src/AlignmentParameterStore.cc [hunk 3]
```diff
@@ -445,53 +446,89 @@ applyAlignableAbsolutePositions( const align::Alignables& alivec,
...
+            std::vector<double> defDiff;
+            for (unsigned int i = 0; i < dold.size(); i++) defDiff.push_back(dnew[i] - dold[i]);
```
Comment (davidlange6) [inline Alignment/CommonAlignmentAlgorithm/src/AlignmentParameterStore.cc]: defDiff.reserve...

File: Alignment/CommonAlignmentAlgorithm/src/AlignmentParameterStore.cc (hunk 1 above)
Comment (usarica) [inline Alignment/CommonAlignmentAlgorithm/src/AlignmentParameterStore.cc]: Why not? The final object propagated is differences, not new old old position.

File: Alignment/CommonAlignmentAlgorithm/src/AlignmentParameterStore.cc (hunk 2 above)
Comment (usarica) [inline Alignment/CommonAlignmentAlgorithm/src/AlignmentParameterStore.cc]: You mean !dold.empty() in place of dol.size()!=0? I can replace that.

File: Alignment/CommonAlignmentAlgorithm/src/AlignmentParameterStore.cc (hunk 3 above)
Comment (usarica) [inline Alignment/CommonAlignmentAlgorithm/src/AlignmentParameterStore.cc]: defDiff.reserve(dold.size())? Sure.

File: Alignment/CommonAlignmentAlgorithm/src/AlignmentParameterStore.cc (hunk 1 above)
Comment (usarica) [inline Alignment/CommonAlignmentAlgorithm/src/AlignmentParameterStore.cc]: Ok, I gues you are referring to the last line. Yes, could be done.

**Comment by usarica** [comment]
The last commit makes changes based on comments by @davidlange6

**Comment by perrotta** [comment]
please test

File: Alignment/CommonAlignmentAlgorithm/src/AlignableDataIORoot.cc [hunk 4]
```diff
@@ -166,11 +170,13 @@ AlignableRelData AlignableDataIORoot::readRelRaw(Alignable* ali,int& ierr)
     pos=pos2;
     rot=rot2;
 
-    // FIXME: Should add reading of deformation values?
-    //        Then call Alignable::setSurfaceDeformation(..) ...
+    for (unsigned int i = 0; i < numDeformationValues_; ++i) {
```
Comment (davidlange6) [inline Alignment/CommonAlignmentAlgorithm/src/AlignableDataIORoot.cc]: hi @usarica - please add deformPars.reserve(deformationValues_.size())
(or maybe better just use a std::copy)

Comment (usarica) [inline Alignment/CommonAlignmentAlgorithm/src/AlignableDataIORoot.cc]: Ok, modified. I am not sure std::copy will be safe, so kept the push_back

**Comment by perrotta** [comment]
please test

**Comment by slava77** [comment]
+1

for https://github.com/cms-sw/cmssw/pull/19584	3bce98c
- updates since the last reco signoff are not related to reco, only in Alignment/CommonAlignmentAlgorithm https://github.com/cms-sw/cmssw/pull/19584/files/7455b7d107deafdc88379b981c90eb4d2bf116dc..3bce98c82067a3ad76e40bf567cdb3e0dd2a0392
- jenkins tests pass

File: Alignment/CommonAlignmentAlgorithm/src/AlignmentParameterStore.cc [hunk 5]
```diff
@@ -445,53 +446,89 @@ applyAlignableAbsolutePositions( const align::Alignables& alivec,
...
+          // If an old surface deformation record exists, ensure that the added deformation has the same type and size.
+          if (!dold.empty() && dtype != SurfaceDeformationFactory::kNoDeformations && dnew.size()==dold.size()){
+            std::vector<double> defDiff; defDiff.reserve(dold.size());
+            for (unsigned int i = 0; i < dold.size(); i++) defDiff.push_back(dnew[i] - dold[i]);
+            auto deform = SurfaceDeformationFactory::create(dtype, defDiff);
+            edm::LogInfo("Alignment") << "@SUB=AlignmentParameterStore::applyAlignableAbsolutePositions"
+              << "Adding surface deformation of type " << SurfaceDeformationFactory::surfaceDeformationTypeName((SurfaceDeformationFactory::Type)deform->type())
+              << ", size " << defDiff.size() << " and first element " << defDiff.at(0)
+              << " to alignable with id / type: " << id << " / " << typeId;
+            ali->addSurfaceDeformation(deform, true);
```
Comment (davidlange6) [inline Alignment/CommonAlignmentAlgorithm/src/AlignmentParameterStore.cc]: would be better than ali took ownership of this pointer so that it didn't have to make a copy of it - but that looks like an old problem...

File: Alignment/HIPAlignmentAlgorithm/plugins/HIPTwoBodyDecayAnalyzer.cc [hunk 6]
```diff
@@ -0,0 +1,385 @@
...
+
+  edm::EDGetTokenT<reco::TrackCollection> alcareco_trackCollToken_;
+  edm::EDGetTokenT<reco::TrackCollection> refit1_trackCollToken_;
+  edm::EDGetTokenT<reco::TrackCollection> ctf_trackCollToken_;
+  edm::EDGetTokenT<reco::TrackCollection> final_trackCollToken_;
+
+  static void fillDescriptions(edm::ConfigurationDescriptions& descriptions);
+
+  TTree* tree;
+  std::vector<std::pair<TString, Float_t*>> floatBranches;
```
Comment (davidlange6) [inline Alignment/HIPAlignmentAlgorithm/plugins/HIPTwoBodyDecayAnalyzer.cc]: please use std::string

File: Alignment/HIPAlignmentAlgorithm/plugins/HIPTwoBodyDecayAnalyzer.cc [hunk 7]
```diff
@@ -0,0 +1,385 @@
...
+  virtual void beginJob() override;
+  virtual void analyze(const edm::Event&, const edm::EventSetup&) override;
+  virtual void endJob() override;
+
+  void bookAllBranches();
+  bool bookBranch(TString bname, TString btype);
+  TString searchArray(TString branchname, int& position);
+  template<typename varType> void setVal(TString bname, varType value){
+    int varposition=-1;
+    TString varbranchtype = searchArray(bname, varposition);
```
Comment (davidlange6) [inline Alignment/HIPAlignmentAlgorithm/plugins/HIPTwoBodyDecayAnalyzer.cc]: could this not just be an enum? would save lots of string parsing it appears

File: Alignment/HIPAlignmentAlgorithm/plugins/HIPTwoBodyDecayAnalyzer.cc [hunk 8]
```diff
@@ -0,0 +1,385 @@
...
+    else if (varbranchtype=="i") *(intBranches.at(varposition).second)=value;
+    else if (varbranchtype=="f") *(floatBranches.at(varposition).second)=value;
+    else std::cerr << "HIPTwoBodyDecayAnalyzer::setVal -> Could not find the type " << varbranchtype << " for branch called " << bname << "!" << std::endl;
+  }
+  void cleanBranches();
+  void initializeBranches();
+  bool actuateBranches();
+
+  void analyzeTrackCollection(
+    TString strTrackType,
```
Comment (davidlange6) [inline Alignment/HIPAlignmentAlgorithm/plugins/HIPTwoBodyDecayAnalyzer.cc]: looks like this is not used?

File: Alignment/HIPAlignmentAlgorithm/plugins/HIPTwoBodyDecayAnalyzer.cc [hunk 9]
```diff
@@ -0,0 +1,385 @@
...
+      return "i";
+    }
+  }
+  for (unsigned int el=0; el<floatBranches.size(); el++){
+    if (branchname==floatBranches.at(el).first){
+      position = el;
+      return "f";
+    }
+  }
+  return "unknown";
```
Comment (davidlange6) [inline Alignment/HIPAlignmentAlgorithm/plugins/HIPTwoBodyDecayAnalyzer.cc]: ok - i can see that indeed, enums would work great here.

File: Alignment/HIPAlignmentAlgorithm/plugins/HIPTwoBodyDecayAnalyzer.cc [hunk 10]
```diff
@@ -0,0 +1,385 @@
...
+    bookBranch(strTrackType + "_MuPlusPt_AfterZVtxFit", "f"); bookBranch(strTrackType + "_MuPlusPz_AfterZVtxFit", "f"); bookBranch(strTrackType + "_MuPlusPhi_AfterZVtxFit", "f");
+    bookBranch(strTrackType + "_ZMass_AfterZVtxFit", "f"); bookBranch(strTrackType + "_ZPt_AfterZVtxFit", "f"); bookBranch(strTrackType + "_ZPz_AfterZVtxFit", "f"); bookBranch(strTrackType + "_ZPhi_AfterZVtxFit", "f");
+    bookBranch(strTrackType + "_MuMinusPt", "f"); bookBranch(strTrackType + "_MuMinusPz", "f"); bookBranch(strTrackType + "_MuMinusPhi", "f");
+    bookBranch(strTrackType + "_MuMinusVertex_x", "f"); bookBranch(strTrackType + "_MuMinusVertex_y", "f"); bookBranch(strTrackType + "_MuMinusVertex_z", "f");
+    bookBranch(strTrackType + "_MuPlusPt", "f"); bookBranch(strTrackType + "_MuPlusPz", "f"); bookBranch(strTrackType + "_MuPlusPhi", "f");
+  }
+  actuateBranches();
+}
+bool HIPTwoBodyDecayAnalyzer::bookBranch(TString bname, TString btype){
+  if (btype=="f") floatBranches.emplace_back(bname, new Float_t);
```
Comment (davidlange6) [inline Alignment/HIPAlignmentAlgorithm/plugins/HIPTwoBodyDecayAnalyzer.cc]: please use float instead of Float_t (here and elsewhere for ints, shorts,etc)

File: Alignment/HIPAlignmentAlgorithm/plugins/HIPTwoBodyDecayAnalyzer.cc (hunk 6 above)
Comment (usarica) [inline Alignment/HIPAlignmentAlgorithm/plugins/HIPTwoBodyDecayAnalyzer.cc]: I could, but this plugin is not really used other than small scale testing. Do we expect TString to be incompatible with CMSSW somehow? Definitely branching is made via const char* or TString, not std::string, so I would have to add .c_str() everywhere else tat calls TTree::Branch.

File: Alignment/HIPAlignmentAlgorithm/plugins/HIPTwoBodyDecayAnalyzer.cc (hunk 7 above)
Comment (usarica) [inline Alignment/HIPAlignmentAlgorithm/plugins/HIPTwoBodyDecayAnalyzer.cc]: I could, but again, this plugin is only for quick tests to make sure TwoBodyDecay classes work properly without crashing (nobody maintains them at this moment, so anything could happen if we don't check).

File: Alignment/HIPAlignmentAlgorithm/python/cdcsTrackSelection_cff_py.txt [hunk 11]
```diff
@@ -0,0 +1,25 @@
+import Alignment.CommonAlignment.tools.trackselectionRefitting as TrackRefitterSequencer
```
Comment (davidlange6) [inline Alignment/HIPAlignmentAlgorithm/python/cdcsTrackSelection_cff_py.txt]: why are these .py.txt files?

File: Alignment/HIPAlignmentAlgorithm/src/HIPAlignmentAlgorithm.cc [hunk 12]
```diff
@@ -743,769 +730,787 @@ void HIPAlignmentAlgorithm::run(const edm::EventSetup& setup, const EventInfo &e
...
+    edm::LogInfo("Alignment") << "@SUB=HIPAlignmentAlgorithm::collector" << "Per-alignable reweighting is turned on. Iterating over the parallel jobs to sum number of hits per alignable for each data type.";
+    // Counting step for per-alignable reweighting
+    for (int ijob=1; ijob<=theCollectorNJobs; ijob++){
+      edm::LogInfo("Alignment") << "@SUB=HIPAlignmentAlgorithm::collector" << "Pre-reading uservar for job " << ijob;
+
+      std::stringstream ss;
+      std::string str;
+      ss << ijob;
+      ss >> str;
+      std::string uvfile = theCollectorPath+"/job"+str+"/"+suvarfilecore;
```
Comment (davidlange6) [inline Alignment/HIPAlignmentAlgorithm/src/HIPAlignmentAlgorithm.cc]: how about 
std::string uvfile = theCollectorPath+"/job"+std::to_string(ijob)+"/"+suvarfilecore;
?

File: Alignment/HIPAlignmentAlgorithm/plugins/HIPTwoBodyDecayAnalyzer.cc (hunk 9 above)
Comment (usarica) [inline Alignment/HIPAlignmentAlgorithm/plugins/HIPTwoBodyDecayAnalyzer.cc]: I agree that the current functionality could be improved a lot. In fact I don't think I need any enums here as long as I have template functions and instantiation done more properly. However, I also think that maybe I can make a separate PR for this instead of delaying this one. Overall we made a lot of improvements that we would also like different people in the Tracker Alignment group to be able to use, but we cannot really put these new functionalities into use coherently without merging the PR. May I ask if you would agree that we open another PR for these changes instead of making this one wait?

File: Alignment/HIPAlignmentAlgorithm/plugins/HIPTwoBodyDecayAnalyzer.cc (hunk 10 above)
Comment (usarica) [inline Alignment/HIPAlignmentAlgorithm/plugins/HIPTwoBodyDecayAnalyzer.cc]: I could do this, but again per my comment above, I would prefer to do this in a separate PR if possible.

File: Alignment/HIPAlignmentAlgorithm/src/HIPAlignmentAlgorithm.cc [hunk 13]
```diff
@@ -743,769 +730,787 @@ void HIPAlignmentAlgorithm::run(const edm::EventSetup& setup, const EventInfo &e
...
+  if (eventtrees->GetSize()>0) theTrackMonitorTree = TTree::MergeTrees(eventtrees);
+  if (hittrees->GetSize()>0) theHitMonitorTree = TTree::MergeTrees(hittrees);
+  // Leave it to HIPAlignmentAlgorithm::terminate to write the trees and close theTrackHitMonitorIORootFile
+
+  delete hittrees;
+  delete eventtrees;
+  for (TFile*& finput : finputlist) finput->Close();
+
+  // Rename the trees to standard names
+  if (theTrackMonitorTree!=0) theTrackMonitorTree->SetName(theTrackMonitorTreeName);
```
Comment (davidlange6) [inline Alignment/HIPAlignmentAlgorithm/src/HIPAlignmentAlgorithm.cc]: please use "nullptr" to pointer comparisons

File: Alignment/HIPAlignmentAlgorithm/python/cdcsTrackSelection_cff_py.txt (hunk 11 above)
Comment (usarica) [inline Alignment/HIPAlignmentAlgorithm/python/cdcsTrackSelection_cff_py.txt]: These are template files that are later replaced inside the actual align_cfg.py of each subjob. This is why their extension was made .txt (even before I was in CMS).

**Comment by davidlange6** [comment]
hi @usarica all - apologies I didn't get all the way through my checks yesterday. I think I did now

File: Alignment/HIPAlignmentAlgorithm/src/HIPAlignmentAlgorithm.cc (hunk 12 above)
Comment (usarica) [inline Alignment/HIPAlignmentAlgorithm/src/HIPAlignmentAlgorithm.cc]: Possible, should I make a separate PR (or add it to the one I proposed) for this instead of making changes in this PR?

**Comment by usarica** [comment]
The PR is updated for the review from @davidlange6 today. Changes have been made for enums in the new testing plugin, removal of TString in favor of std::string, replacement of pointer comparisons/assignments to 0 with nullptr (everywhere in HipPy code). There is no easy solution to not having py.txt files; how HipPy is organized has to change to make that happen, and this is a major work, which I do not think is doable at this time.

**Comment by slava77** [comment]
@cmsbuild please test

**Comment by slava77** [comment]
+1

for https://github.com/cms-sw/cmssw/pull/19584	4e37945
- updates since the last reco signoff are not related to reco, only in Alignment/HIPAlignmentAlgorithm https://github.com/cms-sw/cmssw/pull/19584/files/3bce98c82067a3ad76e40bf567cdb3e0dd2a0392..4e37945e053badab6cbdfaf7c23758ca5a279b19
- jenkins tests pass

**Comment by ianna** [comment]
+1

**Comment by usarica** [comment]
Hi all,
I am planning to open another PR with additional functionalities in HIPAlignmentalgorithm, and a simple bugfix to VertexConstraintProducer (GetByLabel -> GetByToken). Does anyone have a time estimate for when this PR can be merged if there are no objections? The changes I made to HIPAligmentalgorithm start from the commits in this PR.

**Comment by arunhep** [comment]
+1

**Comment by davidlange6** [comment]
+1
